stdlib/domain.ml

```diff
@@ -9,6 +9,61 @@ module Raw = struct
     = "caml_ml_domain_cpu_relax"
 end
 
+module DLS = struct
+
+  type dls_state = Obj.t array
+
+  let unique_value = Obj.repr (ref 0)
+
+  external get_dls_state : unit -> dls_state = "%dls_get"
+
+  external set_dls_state : dls_state -> unit =
+    "caml_domain_dls_set" [@@noalloc]
+
+  let create_dls () =
+    let st = Array.make 8 unique_value in
+    set_dls_state st
+
+  let _ = create_dls ()
+
+  type 'a key = int * (unit -> 'a)
+
+  let key_counter = Atomic.make 0
+
+  let new_key f =
+    let k = Atomic.fetch_and_add key_counter 1 in
+    (k, f)
+
+  let rec log2 n =
+    if n <= 1 then 0 else 1 + (log2 (n asr 1))
+
+  (* If necessary, grow the current domain's local state array such that [idx]
+   * is a valid index in the array. *)
+  let maybe_grow idx =
+    let st = get_dls_state () in
+    if idx < Array.length st then st
+    else begin
+      let sz = Int.shift_left 1 (1 + log2 idx) in
+      let new_st = Array.make sz unique_value in
+      Array.blit st 0 new_st 0 (Array.length st);
+      set_dls_state new_st;
+      new_st
+    end
```

Comment:
The more common way to write this dynamic-growth logic is to double the current size until we get above the target index. I think this more common version is better because (1) it's obvious that the resulting size is above the target index, and (2) in the common case (which is the case here) where the target index is right after the size limit, the new size is computed in O(1) instead of O(log n). Of course, this does not matter very much here, as this operation is infrequent.

Reply:
Fixed.